File: Python/198_HouseRobber.py
class Solution(object):


    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        memo = [-1 for _ in range(len(nums))]

        print(memo)

        def bt(nums, n, memo):
            if n < 0: return 0
            if memo[n] != -1: return memo[n]

            pick = nums[n] + bt(nums,n-2,memo)
            notPick = bt(nums,n-1,memo)

            memo[n] = max(pick, notPick)

            return memo[n]

        return bt(nums,len(nums)-1,memo)

File: C++/198_HouseRobber.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int backtruc(vector<int>& nums, int cur, vector<int>& memo){
        if(cur >= nums.size()) return 0;
        if (memo[cur] != -1) return memo[cur];

        memo[cur] = max(nums[cur] + backtruc(nums, cur+2, memo),  backtruc(nums, cur+1, memo));
        return memo[cur];
    }

    int rob(vector<int>& nums) {
        vector<int> memo(nums.size(),-1);

        return rec(nums, 0, memo);        
    }
};
